### ISSUE TITLE
[engine] scrolling perfomance regression with merged platform/UI thread.

### ISSUE THREAD

**Issue by NabilaWorks** [flutter/flutter#157827]
### Steps to reproduce

1. `flutter channel master`
2. `flutter run --profile`
3. Scroll vigorously up and down

Reproducible on `master` TOT at the moment : 3ed40f003a92487dfa96833e6a1e1bd7f8ceb677 ❌

The `master` commit that caused it : 81e418dd20f784cae63a5eb9bc234b1332f8e925 ❌

Reproducible on `beta` as well 2e2c358c9b14765c90343af9df11e12c5dfc3e6c ❌

The issue got merged here : https://github.com/flutter/flutter/pull/154020

Last good `master` commit : a4b0d973fb196ba86da5c6ae4d51db40a78a0926 ✅

Not reproducible on `stable`  `3.24.4` ✅

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: SingleChildScrollView(
          physics: const BouncingScrollPhysics(parent: AlwaysScrollableScrollPhysics()),
          scrollDirection: Axis.vertical,
          child: Stack(
            children: [
              const SizedBox(
                height: 1440,
                width: 1000,
              ),
              for (var i = 0; i < 100; i++)
                Positioned.fill(
                  top: i * 20,
                  child: Align(
                    alignment: Alignment.topCenter,
                    child: Container(
                      color: Colors.red,
                      height: 15,
                      child: Center(
                        child: Text(
                          '$i',
                          style: const TextStyle(
                            color: Colors.white,
                            fontSize: 11,
                          ),
                        ),
                      ),
                    ),
                  ),
                ),
            ],
          ),
        ),
      ),
    );
  }
}
```

</details>

### Performance profiling on master channel

- [X] The issue still persists on the master channel

### Timeline Traces

<details open><summary>Timeline Traces JSON</summary>

[dart_devtools_2024-10-29_21_26_05.195.json](https://github.com/user-attachments/files/17562727/dart_devtools_2024-10-29_21_26_05.195.json)

</details>

### Video demonstration

<details open>
<summary>Video demonstration</summary>

https://github.com/user-attachments/assets/9c66fca2-762e-40fe-bd2b-0f66bb7593f2

https://github.com/user-attachments/assets/b6fa38ef-e136-4834-9428-ef763e5b9432

https://github.com/user-attachments/assets/27eca780-5428-4aeb-bd27-0dd22d6a7947

</details>

### What target platforms are you seeing this bug on?

iOS

### OS/Browser name and version | Device information

`iOS 18.1 (22B83) | iPhone 12 mini`

### Does the problem occur on emulator/simulator as well as on physical devices?

Unknown

### Is the problem only reproducible with Impeller?

No

### Logs

<details open><summary>Logs</summary>

```console
flutter create scrolling
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.27.0-1.0.pre.292, on macOS 15.1 24B82 darwin-arm64, locale en-US)
    • Flutter version 3.27.0-1.0.pre.292 on channel master at /Users/nabila/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3ed40f003a (14 minutes ago), 2024-10-29 15:45:22 -0400
    • Engine revision 795b5492f1
    • Dart version 3.7.0 (build 3.7.0-78.0.dev)
    • DevTools version 2.40.1
```

</details>

**Comment by jonahwilliams** [comment]
@NabilaWorks please don't `@` contributos in your issues templates, we'll look at it in issue triage normally and that isn't necessary.

**Comment by jonahwilliams** [comment]
I can reproduce this and it seems to be caused by changes to the UI/platform thread - as an aside, we plan to disable this for the next stable release and re-introduce it with an opt out flag for next stable.

I don't see anything that isn't running proportionally slower, nor do i see any big stalls, so i suspect the change to platform/ui thread is causing thread priority/affinity reshuffle and so the raster thread is running on a slower core perhaps. I don't see this regression on any of our benchmarks on CI confusingly enough.

**Comment by jonahwilliams** [comment]
Real devices, iPhone 11. I suspect that if its threading/affinity changes ... then its possible that a more complex app causes the OS to give more priority to raster workloads. This may be an artifact of the app being overly simple

**Comment by jonahwilliams** [comment]
It only seems to be a noticable difference when I scroll more slowly with finger down. Flings seem to behave about the same.

**Comment by jonahwilliams** [comment]
If its related to touch event handling that would also be a reason for automated benchmarks to miss it...

**Comment by jonahwilliams** [comment]
actually, now I feel like I'm going crazy and can't repro it. Will try again tomorrow.

**Comment by jonahwilliams** [comment]
I'm back, well rested, and can repro the problem.

Interestingly enough the issue disappears in release mode, that is if I add `showPerformanceOverlay: true,` to the MaterialApp and run in release mode I don't see any difference in raster time.

For now I plan to add back the opt out but I'm still leaning towards this not being a real regression.

**Comment by jonahwilliams** [comment]
There are lots of differences between profile/release mode unfortunately. Running the VMSevice debugger/profiler has substantial overhead.

**Comment by TheLastFlame** [comment]
I'm not sure if this is related.
But I notice a very strong visual performance sag when using impeller on android on the master branch for the last 6 months or more, even though everything was fine in March.
I don't see this difference in the displayed flutter graphics (it says max frame time 3ms), but at the same time on a device with 120ghz fps seems very low. (when using Skia everything is fine)
I'll think of a better way to record this problem and come back with this. 
But it sounds very similar to what has been described.

**Comment by jonahwilliams** [comment]
That does not sound related at all

**Comment by knopp** [comment]
Is this still relevant? With metal validation disabled I could not reproduce any difference in release mode on a 13pro between `FLTEnableMergedPlatformUIThread` disabled or enabled, both are taking roughly 2.7ms per frame on UI thread.

With metal validation enabled unmerged thread do perform better, but I think that might be caused by the CPU ramping up.

**Comment by jonahwilliams** [comment]
Yeah I think merging the UI/Platform thread changed the threading affinity which made the overhead of metal validation more apparent. I do think this is a slight perf difference as we're now running things like touch input process on the same thread, but worth the cost.
